pkg/compute/assemblyscript.go

```diff
@@ -0,0 +1,230 @@
+			Remediation: fmt.Sprintf("To fix this error, run the following command:\n\n\t$ %s", text.Bold("npm init")),
+		}
+	}
+
+	fmt.Fprintf(out, "Found package.json at %s\n", fpath)
+
+	// 3) Check if `asc` is installed.
+	//
+	// asc is the AssemblyScript compiler. We first check if it exists in the
+	// package.json and then whether the binary exists in the npm bin dirctory.
+	fmt.Fprintf(out, "Checking if AssemblyScript is installed...\n")
+	if !checkPackageDependencyExists("assemblyscript") {
+		return errors.RemediationError{
+			Inner:       fmt.Errorf("`assemblyscript` not found in package.json"),
+			Remediation: fmt.Sprintf("To fix this error, run the following command:\n\n\t$ %s", text.Bold("npm install --save-dev assemblyscript")),
+		}
+	}
+
+	p, err = getNpmBinPath()
+	if err != nil {
+		return errors.RemediationError{
+			Inner:       fmt.Errorf("could not determine npm bin path"),
+			Remediation: fmt.Sprintf("To fix this error, run the following command:\n\n\t$ %s", text.Bold("npm install --global npm@latest")),
+		}
+	}
+
+	path, err := exec.LookPath(filepath.Join(p, "asc"))
+	if err != nil {
+		return fmt.Errorf("getting asc path: %w", err)
+	}
+	if !common.FileExists(path) {
+		return errors.RemediationError{
+			Inner:       fmt.Errorf("`asc` binary not found in %s", p),
+			Remediation: fmt.Sprintf("To fix this error, run the following command:\n\n\t$ %s", text.Bold("npm install --save-dev assemblyscript")),
+		}
+	}
+
+	fmt.Fprintf(out, "Found asc at %s\n", path)
+
+	return nil
+}
+
+// Initialize implments the Toolchain interface and initializes a newly cloned
+// package by installing required dependencies.
+func (a AssemblyScript) Initialize(out io.Writer) error {
+	// 1) Check `npm` is on $PATH
+	//
+	// npm is Node/AssemblyScript's toolchain package manager, it is needed to
+	// install the package dependencies on initialization. We only check whether
+	// the binary exists on the users $PATH and error with installation help text.
+	fmt.Fprintf(out, "Checking if npm is installed...\n")
+
+	p, err := exec.LookPath("npm")
+	if err != nil {
+		return errors.RemediationError{
+			Inner:       fmt.Errorf("`npm` not found in $PATH"),
+			Remediation: fmt.Sprintf("To fix this error, install Node.sj and npm by visiting:\n\n\t$ %s", text.Bold("https://nodejs.org/")),
+		}
+	}
+
+	fmt.Fprintf(out, "Found npm at %s\n", p)
+
+	// 2) Check package.json file exists in $PWD
+	//
+	// A valid npm package manifest file is needed for the install command to
+	// work. Therefore, we first assert whether one exists in the current $PWD.
+	fpath, err := filepath.Abs("package.json")
+	if err != nil {
+		return fmt.Errorf("getting package.json path: %w", err)
+	}
+
+	if !common.FileExists(fpath) {
+		return errors.RemediationError{
+			Inner:       fmt.Errorf("package.json not found"),
+			Remediation: fmt.Sprintf("To fix this error, run the following command:\n\n\t$ %s", text.Bold("npm init")),
+		}
+	}
```

Comment:
suggestion - What about pulling this chunk out into a separate function that `Initialize` and `Verify` could use?

This would also reduce having to update error messages, etc. in two places and make this logic easier to test.

Reply:
I went back and forwards between extracting the shared check in the AS Initialize() and Verify() methods. However opted for the duplication to allow them to remain completely independent of each other, as they’re called by different CLI commands.